src/s_inter.c

```diff
@@ -138,6 +171,472 @@ void sys_set_temppath(void);
+
+static void rbskt_free(t_rbskt* x) {
+    if(x) {
+        rb_free(x->rs_rb);
+        freebytes(x, sizeof(*x));
+    }
+}
+
+static int rbskt_ready(t_rbskt* x) {
+    return x && x->rs_rb && (x->rs_closed || x->rs_errno || rb_available_to_read(x->rs_rb));
+}
+
+int rbskt_bytes_available(t_rbskt* rbskt) {
+    return rb_available_to_read(rbskt->rs_rb);
+}
+
+// to ensure the correct order is preserved when calling send()/sendto()
+// directly from outside rb_dosendone(), always flush the corresponding
+// ringbuffer first
+// should_lock should be set to true if this is not called from the IO thread
+static void sendrb_flush(ring_buffer* rb, int should_lock) {
+    if(should_lock)
+        sys_lockio();
+    rb_dosend(rb, -1);
+    if(should_lock)
+        sys_unlockio();
+}
+
+// A sendto() that writes to a ring_buffer and is real-time safe
+static ssize_t rb_sendto(ring_buffer* rb, int socket, const void *buffer, size_t length, int flags, void* addr, size_t addrlen)
+{
+    //printf("rb_sendto: fd: %d, rb: %p \"%s\"\n", socket, rb, (char*)buffer);
+    size_t maxLength = rb_available_to_write(rb);
+    size_t desiredLength = sizeof(size_t) + length;
+    size_t actualLength = desiredLength <= maxLength ? desiredLength : maxLength;
+    if(actualLength < desiredLength)
+    {
+        perror("rb buffer is full, sending packet from the audio thread");
+        sendrb_flush(rb, 1);
+        // then send out the new data
+        return sendto(socket, buffer, length, 0, addr, addrlen);
+    }
+    rb_write_to_buffer(rb, 5,
+            (char*)&socket, sizeof(socket),
+            (char*)&length, sizeof(length),
+            buffer, length,
+            (char*)&addrlen, sizeof(addrlen),
+            addr, addrlen
+        );
+    return actualLength;
+}
+
+// A send() that writes to a ring_buffer and is real-time safe
+ssize_t rb_send(ring_buffer* rb, int socket, const void *buffer, size_t length, int flags)
+{
+    return rb_sendto(rb, socket, buffer, length, flags, NULL, 0);
+}
+
+// a drop-in replacement for sendto(), exposed for public use, safe to call from a real-time thread
+ssize_t sys_sendto(int sockfd, const void *buf, size_t len, int flags, void* addr, size_t addrlen)
+{
+    if(rbsend_init())
+        return -1;
+    return rb_sendto(pd_this->pd_inter->i_rbsend, sockfd, buf, len, flags, addr, addrlen);
+}
+
+static void gui_failed(const char* s);
+// Read one message from the ring buffer and send it to the network
+static int rb_dosendone(ring_buffer* rb, const int ignoreSigFd)
+{
+    int ret;
+    int socket;
+    size_t length;
+    char* buf;
+    size_t addrlen;
+    struct sockaddr rbaddr;
+    struct sockaddr* addr;
+    if(rb_read_from_buffer(rb, (char*)&socket, sizeof(socket)) < 0)
+    {
+        // no message to retrieve
+        return 0;
+    }
+    if(rb_read_from_buffer(rb, (char*)&length, sizeof(length)) < 0)
+    {
+        perror("we should not be here 1");
+        return -1;
+    }
+    buf = (char*)malloc(length);
```

Comment:
Instead of repeated malloc/free, better use a pre-allocated buffer. Either allocate 65536 bytes (max. UDP packet size), or start with something smaller and grow as needed.

Reply:
ok thanks

Reply:
d15fea5